### ISSUE TITLE
Provide better feedback in recursive custom deserializer error

### ISSUE THREAD

**Issue by fandreuz** [FasterXML/jackson-databind#4554]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

I'm writing a custom deserializer for recursive types, all implementing from a common interface. Some implementations are basically delegates with additional functionalities, see the snippet below for more details. I'm using `@JsonUnwrapped` in combination with Lombok's `@Jacksonized` to make this work.

After a long debugging session I discovered that `JsonUnwrapped` is only considered if the method `unwrappingDeserializer` of the deserializer is overridden, and returns an instance of `JsonDeserializer` which is not the same as the one which is handling the deserialization of the enclosing object -- is there a reason for this check?)

### Reproduction

```java
    interface X {

    }

    @Builder
    @Jacksonized
    @Value
    static class A implements X {
        @JsonUnwrapped
        X value;
    }

    @Builder
    @Jacksonized
    @Value
    static class B implements X {
        String content;
    }

    static class ADeserializer extends StdDeserializer<X> {
        protected ADeserializer() {
            super(X.class);
        }

        @Override
        public X deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
            JsonNode node = p.getCodec().readTree(p);
            if (ctxt.getAttribute("inner") != null) {
                return ctxt.readTreeAsValue(node, B.class);
            }
            ctxt.setAttribute("inner", true);
            return ctxt.readTreeAsValue(node, A.class);
        }
    }

    public static void main(String[] args) throws JsonProcessingException {
        String aJson = "{\"content\" : \"hello\"}";
        SimpleModule module = new SimpleModule();
        module.addDeserializer(X.class, new ADeserializer());
        System.out.println(new ObjectMapper().registerModule(module).readValue(aJson, X.class));
    }
```

### Expected behavior

I would expect this snippet to fail with an explanatory exception pointing at the solution.

**Comment by cowtowncoder** [comment]
I am not quite sure what could be done here: yes, if a deserializer is to support `@JsonUnwrapped`, it must implement that method, to handle the case. I do not know what `@Jacksonized` does, that'd be interesting to know.
But the exception you get is very general one; a deserializer encountered property it does know anything about. It cannot really assume much about behavior of deserializers it delegates to (deserializers know relatively little about each other).

It is currently legal for deserializers to omit definition of `unwrappingDeserializer()`. I guess default implementation could, instead of returning `this`, throw `InvalidDefinitionException`.
But this would be backwards-incompatible change so probably not safe; most likely this could only be done for Jackson 3.0 (`master` branch) -- but that is doable.

**Comment by fandreuz** [comment]
Yeah I had the impression that there was something similar going on..

Why can't we reuse the same instance for the field annotated with `@JsonUnwrapped`? I.e. why do we need the `unwrapping != orig` check [here](https://github.com/joansmith/jackson-databind/blob/230a19d527df7dc0c05017429bcb7aa1cbbe41ef/src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerBase.java#L754-L759) ?

https://github.com/joansmith/jackson-databind/blob/230a19d527df7dc0c05017429bcb7aa1cbbe41ef/src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerBase.java#L754-L759

**Comment by JooHyukKim** [comment]
Maybe because what the comment right below the check says...? 🤔

    // might be cleaner to create new instance; but difficult to do reliably, so:

Also, as per IntelliJ's Github `show history for selection` feature, seems like the check was written by @cowtowncoder  in 2017, 6 years ago. So naturally, there is only a little chance of him remembering it, unless documented/commented.

<img width="1682" alt="image" src="https://github.com/FasterXML/jackson-databind/assets/61615301/c8ef6426-01e7-44c7-a1bd-95268754bbef">

**Comment by JooHyukKim** [comment]
Also, as per IntelliJ's Github `show history for selection` feature, seems like the check was written by @cowtowncoder  in 2017, 6 years ago. So naturally, there is only a little chance of him remembering it, unless documented/commented.

<img width="1682" alt="image" src="https://github.com/FasterXML/jackson-databind/assets/61615301/c8ef6426-01e7-44c7-a1bd-95268754bbef">

**Comment by fandreuz** [comment]
> Maybe because what the comment right below the check says...? 🤔
> 
>     // might be cleaner to create new instance; but difficult to do reliably, so:

Still this does not explain the rationale behind the check... Why can't we use the same instance?

**Comment by cowtowncoder** [comment]
First: the comment referenced no longer exists in 2.18 codebase so I can't say for sure, but it seems misleading. Usually new instances are created and used to allow `JsonDeserializer`s (and most handlers) to be stateless and (for most practical purposes) immutable
(basically, `JsonDeserializer.resolve()` must be allowed to make modifications to handle cyclic definitions but that's about it). 
So the pattern of calling "xxx.withYyy()" means to either return instance as-is if no modifications were needed (property already had value specified), or a newly constructed instance.

Comment seems... odd. Can't explain what it meant.

**Comment by cowtowncoder** [comment]
Not sure what could be done here, closing.

May be re-opened with more info on desired changes.
